kafka/integration_test.go

```diff
@@ -174,39 +178,63 @@ func handleTestEvent(c *Consumer, mt *msgtracker, expCnt int, ev Event) bool {
 		if e.TopicPartition.Error != nil {
 			mt.t.Errorf("Error: %v", e.TopicPartition)
 		}
-		mt.msgs[mt.msgcnt] = e
-		mt.msgcnt++
-		if mt.msgcnt >= int64(expCnt) {
-			return false
+		if mt.msgcnt == int64(cap(mt.msgs)) {
+			mt.t.Logf("***** Buffer full. not saving message: %v\n", e)
+		} else {
+			// mt.t.Logf("Saving message to mt.msgs[%d]: %v\n", mt.msgcnt, e)
+			mt.msgs[mt.msgcnt] = e
+			mt.msgcnt++
 		}
 	case PartitionEOF:
 		break // silence
+	case *Stats:
+		mt.t.Logf("Stats: %v", e)
+		mt.statscnt++
 	default:
+		// there are other unhandled events such as OffsetsCommitted
 		mt.t.Fatalf("Consumer error: %v", e)
+		//mt.t.Logf("Ignored event: %v\n", e)
 	}
+
+	if mt.msgcnt >= int64(expCnt) {
+		if mt.withStats && mt.statscnt == 0 {
+			mt.t.Logf("received message count: %d, wait for stats to show up\n", mt.msgcnt)
+			return true // do not exit yet
+		}
+
+		if mt.withStats {
+			mt.t.Logf("received message count: %d, stats count: %d\n", mt.msgcnt, mt.statscnt)
+		}
+		return false
+	}
+
 	return true
 
 }
 
 // delivery event handler. Tracks the message received
 func deliveryTestHandler(t *testing.T, expCnt int64, deliveryChan chan Event, mt *msgtracker, doneChan chan int64) {
-
 	for ev := range deliveryChan {
-		m, ok := ev.(*Message)
-		if !ok {
-			continue
-		}
+		switch e := ev.(type) {
```

Comment:
Yes,  the following two tests show stats events ended on delivery channel:
go test -v -run TestProducerChannelStats
go test -v -run TestProducerFuncStats